controller/src/main/java/io/pravega/controller/eventProcessor/impl/EventProcessorGroupImpl.java

```diff
@@ -262,7 +284,66 @@ public void changeEventProcessorCount(int count) throws CheckpointStoreException
     public Set<String> getProcesses() throws CheckpointStoreException {
         return checkpointStore.getProcesses();
     }
+    
+    @VisibleForTesting
+    void rebalance() {
+        try {
+            ReaderSegmentDistribution readerSegmentDistribution = readerGroup.getReaderSegmentDistribution();
+            Map<String, Integer> distribution = readerSegmentDistribution.getReaderSegmentDistribution();
+            int readerCount = distribution.size();
+            int unassigned = readerSegmentDistribution.getUnassignedSegments();
+            int segmentCount = distribution.values().stream().reduce(0, (x, y) -> x + y) + unassigned;
+            
+            // If there are idle readers (no segment assignments, then identify and replace overloaded readers). 
+            boolean idleReaders = distribution.entrySet().stream().anyMatch(x -> !Strings.isNullOrEmpty(x.getKey()) && x.getValue() == 0);
+            if (idleReaders) {
+                distribution.forEach((readerId, assigned) -> {
+                    if (!Strings.isNullOrEmpty(readerId)) {
+                        // check if the reader belongs to this group and the reader is eligible for rebalance
+                        if (eventProcessorMap.containsKey(readerId) && isRebalanceCandidate(assigned, readerCount, segmentCount)) {
+                            replaceCell(readerId);
+                        }
+                    }
+                });
+            }
+        } catch (Exception e) {
+            Throwable realException = Exceptions.unwrap(e);
+            log.warn("Rebalance failed with exception {} {}", realException.getClass().getSimpleName(), e.getMessage());
+        }
+    }
 
+    private boolean isRebalanceCandidate(int assigned, int readerCount, int segmentCount) {
+        double fair = (double) segmentCount / (double) readerCount;
```

Comment:
The segment count also includes `unassigned` segments. These unassigned segments would be eventually distributed across the existing readers. Shouldn't `isRebalanceCandidate()` be based on the unassigned segments so as to prevent unnecessary rebalance operations?

Reply:
The objective is to rebalance load from any overloaded reader but to not do it too aggressively either. 
So we are rebalancing only if a reader is overloaded AND there are idle readers in the group. 
If such imbalance is present its an opportunity to rebalance. 
Idle readers and unassigned segments is an interesting case where one could argue that idle readers can be picking up unassigned segments and may no longer be idle. But we still have an overloaded reader and we treat it as an opportunity to rebalance the overloaded reader.